oak_runtime/src/node/http/user.rs

```diff
@@ -0,0 +1,303 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! A [`UserNode`] is a virtual node that is created per HTTP request and serves as a router between
+//! an HTTP server pseudo-Node and an Oak node that serves the incoming HTTP requests. The
+//! [`UserNode`] forwards the requests to the Oak node, and returns the responses back to the HTTP
+//! server pseudo-Node. We set the user's identity (extracted from the incoming request) as both the
+//! declassification and endorsement privileges of the [`UserNode`]. This allows the [`UserNode`] to
+//! assign user-specific labels to the communication channels with the Oak node. The HTTP server
+//! pseudo-Node cannot create channels with such user-specific labels itself. User-specific labels
+//! are required to prevent any potential data leaks through the Oak node.
+
+use crate::{
+    io::{Receiver, ReceiverExt, Sender, SenderExt},
+    node::{ConfigurationError, Node, NodePrivilege},
+    proto::oak::invocation::{HttpInvocation, OuterHttpInvocation},
+    RuntimeProxy,
+};
+use log::{error, trace, warn};
+use oak_abi::{
+    label::{Label, Tag, UserIdentityTag},
+    proto::oak::application::UserNodeConfiguration,
+};
+use oak_io::handle::{ReadHandle, WriteHandle};
+use oak_services::proto::oak::encap::HttpRequest;
+use std::collections::HashSet;
+use tokio::sync::oneshot;
+
+/// Struct that represents a user node, with its privilege set to the user's identity as both the
+/// declassification and the endorsement tags.
+pub struct UserNode {
+    /// Pseudo-Node name.
+    node_name: String,
+    /// The privilege of the node.
+    privilege: NodePrivilege,
+}
+
+impl UserNode {
+    pub fn new(node_name: &str, config: UserNodeConfiguration) -> Result<Self, ConfigurationError> {
+        let mut privilege_tags = HashSet::new();
+        let tag = match config.privilege.is_empty() {
+            true => None,
+            false => Some(oak_abi::label::tag::Tag::UserIdentityTag(UserIdentityTag {
+                public_key: config.privilege,
```

Comment:
The naming seems confusing: using `privilege` in the proro to represent a public key in the config, but later it is used to represent the entire node privilege.